### PR TITLE
Simplify Mongo Client integration with CDI

### PR DESCRIPTION
**Pull request by geoand**

### CODE REVIEW

**Comment by geoand** [comment]
@gsmet @cescoffier @loicmathieu FYI

**Comment by geoand** [comment]
CI Failure is:

```
[INFO] Quarkus - Integration Tests - OpenID Connect Adapter Code Flow FAILURE [ 46.428 s]
```

I will however restart the tests since in that group the Mongo tests were not run, so let's be extra careful and make sure they run.

**Comment by loicmathieu** [comment]
@geoand can you explain in a few line what you try to achieve with this PR ?
To help me understand it ;)

**Comment by geoand** [comment]
@loicmathieu sure thing!

The are two things going:

* Remove all the ugly volative fields configuration and replace with it with better integration with CDI.
* Remove all manually generated bytecode and replace it with Arc synthetic beans.

We did the same exercise with Agroal Datasource handling and it makes the code much more maintainable.

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java [hunk 1]
```diff
@@ -119,233 +103,105 @@ public void mongoClientNames(ApplicationArchivesBuildItem applicationArchivesBui
...
+            // named reactive client
+            syntheticBeanBuildItemBuildProducer
+                    .produce(createSyntheticBean(recorder, mongodbConfig, makeUnremovable, mongoClientName.getName(), true));
+        }
+    }
 
-        List<ConnectionPoolListener> poolListenerList = connectionPoolListenerProvider.stream()
-                .map(MongoConnectionPoolListenerBuildItem::getConnectionPoolListener)
-                .collect(Collectors.toList());
+    private SyntheticBeanBuildItem createSyntheticBean(MongoClientRecorder recorder, MongodbConfig mongodbConfig,
```
Comment (loicmathieu) [inline extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:177]: I don't like the boolean to switch from the sync and reactive client. I found it not really readable.
Can we have two methods instead ?

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java [hunk 2]
```diff
@@ -10,44 +11,44 @@
...
+            }
+        };
+    }
 
-    public BeanContainerListener addMongoClient(
-            Class<? extends AbstractMongoClientProducer> mongoClientProducerClass,
-            boolean disableSslSupport) {
-        return new BeanContainerListener() {
+    public Supplier<MongoClient> mongoClientSupplier(String clientName,
+            @SuppressWarnings("usused") MongodbConfig mongodbConfig) {
```
Comment (loicmathieu) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:32]: If it's unsed, why keeping it ?

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java [hunk 3]
```diff
@@ -10,44 +11,44 @@
...
-        // that were not properly injected but at least the config objects present in
-        // the map)
-        // The elements from the default mongoClient are there
-        AbstractMongoClientProducer producer = Arc.container().instance(AbstractMongoClientProducer.class).get();
-        producer.setCodecs(codecs);
-        producer.setBsonDiscriminators(bsonDiscriminators);
-        producer.setConfig(config);
-        producer.setConnectionPoolListeners(connectionPoolListeners);
+    public Supplier<ReactiveMongoClient> reactiveMongoClientSupplier(String clientName,
+            @SuppressWarnings("usused") MongodbConfig mongodbConfig) {
```
Comment (loicmathieu) [inline extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:43]: If it's unsed, why keeping it ?

**Review by loicmathieu — CHANGES_REQUESTED** [review]
This seems OK for me but I'm not on expert of the build system ...
I made some small remarks / questions

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java (hunk 2 above)
Comment (geoand) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:32]: We need it to make sure that this recorder method is executed after the config system has been setup - see the comment in MongoProcessor where `mongoClientSupplier` is called

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java (hunk 1 above)
Comment (geoand) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java:177]: Sure, I was on the fence about this myself.

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java (hunk 3 above)
Comment (geoand) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:43]: See my comment above

File: extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java (hunk 1 above)
Comment (geoand) [reply extensions/mongodb-client/deployment/src/main/java/io/quarkus/mongodb/deployment/MongoClientProcessor.java]: Fixed

File: extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java (hunk 2 above)
Comment (loicmathieu) [reply extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/MongoClientRecorder.java:32]: OK, the fact that the comment is on the call site and not the parameter declaration site make me miss that point.

**Review by loicmathieu — APPROVED** [review]
LGTM

**Review by gsmet — APPROVED** [review]

**Pull request merged**